torchvision/transforms/_core.py

```diff
@@ -0,0 +1,83 @@
+import collections.abc
+from typing import Any, Callable, Optional, Set, Tuple, Type, TypeVar
+
+import torch
+from torch import nn
+
+from torchvision.features import BoundingBox, Feature, Image
+
+T = TypeVar("T")
+
+
+__all__ = ["Transform"]
+
+
+class Transform(nn.Module):
+    _DEFAULT_FEATURE_TYPE: Type[Feature]
+
+    def _apply_feature_transforms(self, sample: Any) -> Any:
+        raise NotImplementedError
+
+    @staticmethod
+    def _apply_to_sample(sample: Any, fn: Callable[[Any], Any]) -> Any:
+        if isinstance(sample, collections.abc.Sequence):
+            return [Transform._apply_to_sample(data, fn) for data in sample]
+        elif isinstance(sample, collections.abc.Mapping):
+            return {name: Transform._apply_to_sample(data, fn) for name, data in sample.items()}
+        else:
+            return fn(sample)
+
+    @staticmethod
+    def _apply_to_feature_type(sample: Any, feature_type: Type[Feature], fn: Callable[[Any], Any]) -> Any:
+        return Transform._apply_to_sample(sample, lambda data: fn(data) if isinstance(data, feature_type) else data)
+
+    @staticmethod
+    def _query_sample(data: Any, fn: Callable[[Any], Optional[T]]) -> Set[T]:
+        if isinstance(data, (collections.abc.Sequence, collections.abc.Mapping)):
+            results = set()
+            for item in data.values() if isinstance(data, collections.abc.Mapping) else data:
+                results.update(Transform._query_sample(item, fn))
+            return results
+        else:
+            result = fn(data)
+            return {result} if result is not None else {}
+
+    @staticmethod
+    def _query_feature_type(sample: Any, feature_type: Type[Feature], fn: Callable[[Any], Optional[T]]) -> Set[T]:
+        return Transform._query_sample(sample, lambda data: fn(data) if isinstance(data, feature_type) else None)
+
+    def forward(self, *inputs):
+        sample = self._apply_to_sample(
+            inputs if len(inputs) > 1 else inputs[0],
+            lambda data: self._DEFAULT_FEATURE_TYPE(data) if type(data) is torch.Tensor else data,
+        )
+        return self._apply_feature_transforms(sample)
+
+    @staticmethod
+    def _apply_image_and_bounding_box_transforms(
```

Comment:
Mentioned in person, I think if we let `image_size` be a property of `BoundingBox` then this is not needed

Reply:
Yes. If we make each `Feature` self-contained, this can be a lot easier.